Gems/AtomTressFX/Code/Rendering/HairCommon.cpp

```diff
@@ -0,0 +1,205 @@
+            //! The association with all shaders will work properly.
+            Data::Instance<RPI::ShaderResourceGroup> UtilityClass::CreateShaderResourceGroup(
+                Data::Instance<RPI::Shader> shader,
+                const char* shaderResourceGroupId,
+                const char* moduleName)
+            {
+                Data::Instance<RPI::ShaderResourceGroup> srg = RPI::ShaderResourceGroup::Create(shader->GetAsset(), AZ::Name{ shaderResourceGroupId });
+                if (!srg)
+                {
+                    AZ_Error(moduleName, false, "Failed to create shader resource group");
+                    return nullptr;
+                }
+                return srg;
+            }
+
+
+            //! If srg is nullptr the index handle will NOT be set.
+            //! This can be useful when creating a constant buffer or an image.
+            Data::Instance<RPI::Buffer> UtilityClass::CreateBuffer(
+                const char* warningHeader,
+                SrgBufferDescriptor& bufferDesc,
+                Data::Instance<RPI::ShaderResourceGroup> srg)
+            {
+                // If srg is provided, match the shader Srg bind index (returned via the descriptor)
+                if (srg)
+                {   // Not always do we want to associate Srg when creating a buffer
+                    bufferDesc.m_resourceShaderIndex = srg->FindShaderInputBufferIndex(bufferDesc.m_paramNameInSrg).GetIndex();
+                    if (bufferDesc.m_resourceShaderIndex == uint32_t(-1))
+                    {
+                        AZ_Error(warningHeader, false, "Failed to find shader input index for [%s] in the SRG.",
+                            bufferDesc.m_paramNameInSrg.GetCStr());
+                        return nullptr;
+                    }
+                }
+
+                // Descriptor setting
+                RPI::CommonBufferDescriptor desc;
+                desc.m_elementFormat = bufferDesc.m_elementFormat;
+                desc.m_poolType = bufferDesc.m_poolType;;
+                desc.m_elementSize = bufferDesc.m_elementSize;
+                desc.m_bufferName = bufferDesc.m_bufferName.GetCStr();
+                desc.m_byteCount = (uint64_t)bufferDesc.m_elementCount * bufferDesc.m_elementSize;
+                desc.m_bufferData = nullptr;    // set during asset load - use Update
+
+                // Buffer creation
+                return RPI::BufferSystemInterface::Get()->CreateBufferFromCommonPool(desc);
+            }
+
+
+            Data::Instance<RPI::Buffer> UtilityClass::CreateBufferAndBindToSrg(
+                const char* warningHeader,
+                SrgBufferDescriptor& bufferDesc,
+                Data::Instance<RPI::ShaderResourceGroup> srg)
+            {
+                // Buffer creation
+                Data::Instance<RPI::Buffer> buffer = CreateBuffer(warningHeader, bufferDesc, srg);
+
+                if (!buffer)
```

Comment:
Agree with Tommy, there's some reduction of code duplication that can be done here. Not a huge deal but would help

Reply:
I understand that you are referring to the fact that I created yet another method called `CreateBufferAndBindToSrg ` that while using the `CreateBuffer`, does not take advantage of `BindBufferToSrg`.
It is a very subtle difference but it can make a world of difference.  Same buffer could have been created and bound (in this case use the method `CreateBufferAndBindToSrg`), or could have been create but bound differently to several Srgs, in this case you'd separate the creation from the binding and can have multiple different bindings.
In the case of multiple bindings / Srgs I would not like to use the index but to reacquire it every time, hence create once but use `BindBufferToSrg`.
Having said that, you are right and at the expense of calling `FindShaderInputBufferIndex` twice, the change should work.  Since it's not called too often I will change and test.